File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

class Main {

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");

        int N = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(st.nextToken());
        int[] lans = new int[N];

        for (int i = 0; i < N; i++) {
            lans[i] = Integer.parseInt(br.readLine());
        }

        sb.append(binary_search(lans, K));
        System.out.println(sb);

    }

    private static long binary_search(int[] lans, int K) {
        long start = 1;
        long end = Arrays.stream(lans).max().getAsInt();

        while (start <= end) {
            long count = 0;
            long mid = (start + end) / 2;
            for (int lan : lans) {
                count += lan / mid;
            }
            if (count >= K) {
                start = mid + 1;
            } else {
                end = mid - 1;
            }
        }

        return  end;
    }


}

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
import sys

N, K = map(int, sys.stdin.readline().split())

lans = [int(sys.stdin.readline().strip()) for _ in range(N)]

# 백준 1654번 문제에서 이분 탐색 쓰는 이유는? : 이분 탐색은 최적화 문제를 해결할 때 사용한다.
# 최적화 문제란? : 최적화 문제는 여러 해답 중 하나를 선택해야 할 때, 그 해답을 찾는 문제이다.
# 최적화 문제의 특징 : 최적화 문제는 문제의 조건을 만족하는 해가 여러 개일 수 있다. 이 때, 이분 탐색을 사용하면 효과적으로 해결할 수 있다.

lans.sort()

def binary_search(lans, K):
    start = 1
    end = lans[-1]

    while start <= end:
        mid = (start + end) // 2
        count = 0

        for lan in lans:
            count += lan // mid

        if count >= K:
            start = mid + 1
        else:
            end = mid - 1

    return end


print(binary_search(lans, K))


